File: tests/SimpleTester.java
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.OutputStreamWriter;
import java.util.Calendar;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

public class SimpleTester extends DefaultHandler {

   private String temp;
   
   /** The main method sets things up for parsing */
   public static void main(String[] args) throws IOException, SAXException, java.io.FileNotFoundException, ParserConfigurationException {
      
      //Create a "parser factory" for creating SAX parsers
      SAXParserFactory spfac = SAXParserFactory.newInstance();
      
      //Now use the parser factory to create a SAXParser object
      SAXParser sp = spfac.newSAXParser();
      
      //Create an instance of this class; it defines all the handler methods
      SimpleTester handler = new SimpleTester();
      
      //Finally, tell the parser to parse the input and notify the handler
      try {
         sp.parse(args[0], handler);
      } catch (Exception fnfe) {
         System.out.println("tried to parse " + args[0] + ". error: " + fnfe.getMessage());
      }
      File myFile = new File(args[0]);
      System.out.println(myFile.getAbsolutePath() + " was last modified " + myFile.lastModified());
      Calendar fileCalendar = Calendar.getInstance();
      fileCalendar.setTimeInMillis(myFile.lastModified());
      System.out.println("which means year: " + fileCalendar.get(Calendar.YEAR) + ", month: " + fileCalendar.get(Calendar.MONTH) + ", date: " + fileCalendar.get(Calendar.DAY_OF_MONTH) + ", hour: " + fileCalendar.get(Calendar.HOUR_OF_DAY) + ", minute: " + fileCalendar.get(Calendar.MINUTE) + ", second: " + fileCalendar.get(Calendar.SECOND));
      File file = new File("/tmp/bestofjava/java_time.txt");
 
      // if file doesnt exists, then create it
      if (!file.exists()) {
         file.createNewFile();
      }
 
      FileWriter fw = new FileWriter(file.getAbsoluteFile());
      BufferedWriter bw = new BufferedWriter(fw);
      bw.write(Long.toString(System.currentTimeMillis()));
      bw.newLine();
      Calendar oldCal=Calendar.getInstance(),newCal=Calendar.getInstance();
      
      if (myFile.exists()) System.out.println("oh yes, " + myFile.getName() + " exists");
      else System.out.println("crap, " + myFile.getName() + " does not exist");
      System.out.println("the absolute path is " + myFile.getAbsolutePath());
      System.out.println("the parent dir is " + myFile.getParent());
      File parentDir = new File(myFile.getParent());
      if (parentDir.isDirectory()) System.out.println(parentDir.getAbsolutePath() + " is a directory");
      else System.out.println(parentDir.getAbsolutePath() + " is not a directory");
      oldCal.setTimeInMillis(myFile.lastModified());
      newCal.setTimeInMillis(System.currentTimeMillis());
      if (oldCal.after(newCal)) System.out.println("oups, " + oldCal.getTimeInMillis() + " is after " + newCal.getTimeInMillis());
      if (newCal.before(oldCal)) System.out.println("oups, " + newCal.getTimeInMillis() + " is before " + oldCal.getTimeInMillis());
      Calendar getset = Calendar.getInstance();
      getset.set(2014, 3, 5);
      bw.write(Long.toString(getset.getTimeInMillis())); //2014-04-05 in millis
      bw.close();
      System.out.println("year: " + getset.get(Calendar.YEAR) + ", month: " + getset.get(Calendar.MONTH) + ", date: " + getset.get(Calendar.DAY_OF_MONTH));
      getset.setTimeInMillis(getset.getTimeInMillis());
      PrintWriter pw = new PrintWriter(new OutputStreamWriter(System.out));
      pw.println("after calendar time is reset in millis: year: " + getset.get(Calendar.YEAR) + ", month: " + getset.get(Calendar.MONTH) + ", date: " + getset.get(Calendar.DAY_OF_MONTH));
      pw.flush();
      pw.close();
   }


   /*
    * When the parser encounters plain text (not XML elements),
    * it calls(this method, which accumulates them in a string buffer
    */
   public void characters(char[] buffer, int start, int length) {
      temp += new String(buffer, start, length);
   }
   

   /*
    * Every time the parser encounters the beginning of a new element,
    * it calls this method, which resets the string buffer
    */ 
   public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
      temp = "";
      System.out.print("<" + qName + ">");
      if (!uri.equals("")) System.out.print("uri: " + uri + " ");
      if (!localName.equals("")) System.out.print("localName: " + localName + " ");
//   for(int i=0; i < attributes.getLength(); i++)
      //  System.out.print(attributes.getQName(i) + " has type " + attributes.getType(i) + " ");
   }
   
   /*
    * When the parser encounters the end of an element, it calls this method
    */
   public void endElement(String uri, String localName, String qName)
                     throws SAXException {
      if (!uri.equals("")) System.out.print("uri: " + uri + " ");
      if (!localName.equals("")) System.out.print("localName: " + localName + " ");
      if (!temp.equals("")) System.out.print("characters: " + temp);
      System.out.println("</" + qName + ">");
   }
   
   
   
}

File: tests/SimpleTester.cpp
/* This hack is used to perform some simple tests such as parse an xml file and
   make some printouts that may be compared with those of a similar java test
   program. 
   Copyright (C) 2014 Martin Nylin

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include <fstream>
#include <iostream>
#include <string>
#include "SimpleTester.hpp"
#include "javax/xml/parsers/SAXParserFactory.hpp"
#include "java/io/PrintWriter.hpp"
#include "java/lang/System.hpp"
#include "java/util/Calendar.hpp"

/*
 * When the parser encounters plain text (not XML elements),
 * it calls(this method, which accumulates them in a string buffer
 */
void SimpleTester::characters(const std::string& chars) {
   temp.append(chars);
}
      

/*
 * Every time the parser encounters the beginning of a new element,
 * it calls this method, which resets the string buffer
 */ 
void SimpleTester::startElement(const std::string& uri,
                                             const std::string& localName,
                                             const std::string& qName,
                                             const bestofjava::Attributes& attributes)  {
   temp = "";
   std::cout << "<" << qName << ">";
   if (!uri.empty()) std::cout << "uri: " << uri << " ";
   if (!localName.empty()) std::cout << "localName: " << localName << " ";
}

/*
 * When the parser encounters the end of an element, it calls this method
 */
void SimpleTester::endElement(const std::string&    uri,
                                           const std::string&   localName,
                                           const std::string&   qName) {
   if (!uri.empty()) std::cout << "uri: " << uri << " ";
   if (!localName.empty()) std::cout << "localName: " << localName << " ";
   if (!temp.empty()) std::cout << "characters: " << temp;
   std::cout << "</" << qName << ">" << std::endl;
}


int main(int argc, char **argv) {
   std::ofstream myfile;
   myfile.open ("/tmp/bestofjava/cpp_time.txt");
   myfile << bestofjava::System::currentTimeMillis() << std::endl;
   myfile.close();
   
   //Create a "parser factory" for creating SAX parsers
   bestofjava::SAXParserFactory spfac = bestofjava::SAXParserFactory::newInstance();
   
   //Now use the parser factory to create a SAXParser object
   bestofjava::SAXParser sp = spfac.newSAXParser();
   
   //Create an instance of this class; it defines all the handler methods
   SimpleTester handler;
   
   //Finally, tell the parser to parse the input and notify the handler
   bestofjava::File myFile(argv[1]);
   sp.parse(myFile, handler);
   std::cout << myFile.getAbsolutePath() << " was last modified " << myFile.lastModified() << std::endl;
   bestofjava::Calendar fileCalendar;
   fileCalendar.setTimeInMillis(myFile.lastModified());
   std::cout << "which means year: " << fileCalendar.get(bestofjava::Calendar::YEAR) << ", month: " << fileCalendar.get(bestofjava::Calendar::MONTH) << ", date: " << fileCalendar.get(bestofjava::Calendar::DAY_OF_MONTH) << ", hour: " << fileCalendar.get(bestofjava::Calendar::HOUR_OF_DAY) << ", minute: " << fileCalendar.get(bestofjava::Calendar::MINUTE) << ", second: " << fileCalendar.get(bestofjava::Calendar::SECOND) << std::endl;
   bestofjava::Calendar oldCal,newCal;
   if (myFile.exists()) std::cout << "oh yes, " << myFile.getName() << " exists" << std::endl;
   else std::cout << "crap, " << myFile.getName() << " does not exist" << std::endl;
   std::cout << "the absolute path is " << myFile.getAbsolutePath() << std::endl;
   std::cout << "the parent dir is " + myFile.getParent() << std::endl;
   bestofjava::File parentDir(myFile.getParent());
   if (parentDir.isDirectory()) std::cout << parentDir.getAbsolutePath() << " is a directory" << std::endl;
   else std::cout << parentDir.getAbsolutePath() << " is not a directory" << std::endl;
   oldCal.setTimeInMillis(myFile.lastModified());
   newCal.setTimeInMillis(bestofjava::System::currentTimeMillis());
   if (oldCal.after(newCal)) std::cout << "oups, " << oldCal.getTimeInMillis() << " is after " << newCal.getTimeInMillis() << std::endl;
   if (newCal.before(oldCal)) std::cout << "oups, " << newCal.getTimeInMillis() << " is before " << oldCal.getTimeInMillis() << std::endl;
   // bestofjava::Calendar getset;
   // int year=2014, month=4, day_of_month=5;
   // getset.set(year, month, day_of_month);
   // std::cout << "2014-04-05 in millis: " << getset.getTimeInMillis() << std::endl;
   // std::cout << "year: " << getset.get(bestofjava::Calendar::YEAR) << ", month: " << getset.get(bestofjava::Calendar::MONTH) << ", date: " << getset.get(bestofjava::Calendar::DAY_OF_MONTH) << ", hour: " << getset.get(bestofjava::Calendar::HOUR_OF_DAY) << ", minute: " << getset.get(bestofjava::Calendar::MINUTE) << ", second: " << getset.get(bestofjava::Calendar::SECOND) << std::endl;
   // getset.setTimeInMillis(getset.getTimeInMillis());
   // std::cout << "after calendar time is reset in millis: year: " << getset.get(bestofjava::Calendar::YEAR) << ", month: " << getset.get(bestofjava::Calendar::MONTH) << ", date: " << getset.get(bestofjava::Calendar::DAY_OF_MONTH) << ", hour: " << getset.get(bestofjava::Calendar::HOUR_OF_DAY) << ", minute: " << getset.get(bestofjava::Calendar::MINUTE) << ", second: " << getset.get(bestofjava::Calendar::SECOND) << std::endl;
   
   
   
}




